### ISSUE TITLE
The codecs parameter should have a formal definition of the use of the combination operators.

### ISSUE THREAD

**Issue by cconcolato** [w3c/tt-profile-registry#71]
As proposed in #63, Section 2 should define formally the use of the + and | operators.

**Comment by cconcolato** [comment]
In particular, the proposal in bullet point 2 of https://github.com/w3c/tt-profile-registry/issues/63#issuecomment-468380910 is a good text.

**Comment by mikedo** [comment]
Copying #63 substance to this new issue so we don't have to cover old ground...

mikedo commented on Feb 22, 2019

> A reminder that any changes to the media type information in section 2 requires IANA and IETF expert review through the normal process between W3C and IANA. This should be done before publication.

>The codecs parameter was created to be used in ISO BMFF and DASH MPD. Thus, it needs to conform to RFC 6381 syntax. That is, the entire codecs parameter string must conform to RFC 6381 (not just the profile code). The RFC 6381 citation should be moved to the codecs parameter text.

>If we want to constrain the charset of the profile code, then we can point to 4CC in ISO/IEC 14496-12. Note that a recent update to that constrained 4CC to a subset of ASCII (0x20-0x7E).

skynavga commented on Feb 28, 2019

>@mikedo a few questions/comments

>Since the W3C is the designated "change controller", then is it correct to say that an IETF expert review is strictly a formality, and not required by process?

>As specified, the registration does not refer to RFC 6381 and does not define a formal syntax (or charset) but implies that the syntax is effectively

>codecs-parameter-value :
  short-identifier [ ( "|" | "+" ) short-identifier ]*

>short-identifier :
  4*<any TOKEN character except "|" and "+">
I can see that it may desirable to make this syntax more concrete, and perhaps even more restricted, e.g., allowing only ASCII alpha and digit characters in a short-identifier.

>Do you think we need to do this in the next edition of the published NOTE or in a subsequent edition (giving us more time to resolve these changes separately from pushing out the door our current changes)?

mikedo commented on Feb 28, 2019

>I'm confident in the review process for changes. Philippe manages it I believe (he did last time).

>RFC 6381 is cited here: https://w3c.github.io/tt-profile-registry/#Registration_Entry_Requirements_and_Update_Process It needs to be in the codecs parameter paragraph instead. It does not help to have it for the individual codes.

>We need to ensure the syntax is valid now hopefully before anyone uses the more complex operator syntax. I have a 6381 regex if someone will prepare a few complex examples, If they pass, then we can defer on perfecting the syntax. The 6381 citation needs to be fixed though so that we're covered until there is more concrete syntax,

skynavga commented on Mar 1, 2019

>Here are some possible values for the codecs parameter

>im2t
im2t|im2i
im2t+im2i
tt2p|tt1p|im1t|im1i
tt2p+tt2t|tt1p+tt1t|im1t|im1i

mikedo commented on Mar 5, 2019

>I have a draft work in process XML regex for DASH MPD @codecs (RFC 6381). The above examples all validate. That’s good and gives some level of confidence that we can defer a more concrete syntax for the profile code itself.

**Comment by mikedo** [comment]
I look forward to a proposal for the IANA codecs parameter syntax (I am not the proponent of this). Once that is baked, then the registry update is trivial.

**Comment by nigelmegitt** [comment]
Thanks for this @mikedo , very helpful.

**Comment by mikedo** [comment]
And from PR #69 :

cconcolato commented on Mar 14, 2019

>I'm not sure the note as is really useful. It's true that the codecs parameter of TTML follows a specific pattern, either a 4CC or a set of +/|-combined 4CCs, but the purpose of this section should not be about how the codecs in TTML is a specific subset of the general RFC6381 one. The note in this section could rather say: "the restriction on the + and | signs is required to enable the use of these signs in the codecs parameter, as defined in the MIME type registration". We could add a note in the MIME type saying: "Note that in this MIME type registration the first element (as defined in RFC6381) of the codecs parameter may contain a combination of 4CC and the + and | signs".

**Comment by nigelmegitt** [comment]
In beginning to prepare a pull request for this I am puzzled that the issue has been raised. The following text is already present in a normative section:

> Additionally, applications using the entries in the registry are encouraged to adopt the following combination syntax:
>
>Employ two combination operators, '+' (AND) and '|' (OR), which may be used to specify, respectively, that multiple processor profiles apply (simultaneously) or that any processor profile of a list of profiles may apply individually. If both operators are used in a codecs value, then the '+' operator has precedence.
>
>The example: "A+B|C+D|E" states that a TTML processor that implements any one of A+B or C+D or E processor profiles satisfies, at first order, the requirements to fetch and begin decode/processing of a TTML document, where X+Y means that both X and Y processor profiles must be supported, and X|Y means that either X or Y processor profile must be supported.

This seems to be exactly what this issue asks for, unless the request is simply to write out the syntax for this in a more formal seeming style, without changing the content. For example:

```
mediatype: "application/ttml+xml" [charset]? [profile]? [codecs]?

charset: [XML encoding]

profile: [...]

codecs: codec-group [ | codec_group ]*

codec-group: 4cc [+ 4cc]*

4cc: 4 character code defined by this specification
```

**Comment by mikedo** [comment]
No harm in a formal ABNF, but that's not the issue that I tried to raise.

This profiles document states:

> An identifier must conform with the **element** non-terminal of [RFC6381] and, furthermore, may not contain any of the characters in the regular expression character class [+|].

> the codecs parameter ... is a syntactic subset of the definition of the codecs parameter defined by [RFC6381].

The syntax defined in this profile document, does not directly have anything to do with RFC 6381, except that the charset is constrained to that of "element", which among other things forbids ".".. The codecs parameter defined here is for media type, "application/ttml+xml".  So, the use of RFC 6381 is a little misleading.

The codecs parameter of **media types used in ISO BMFF** (specifically media type application/mp4) was created by RFC 6381 and extended by other media specifications such as 14496-30. The latter, following RFC 6381, states that the codecs parameter for a TTML document carried in ISOI BMFF has the syntax "stpp.ttml" appended with an optional profile syntax citing this W3C document (including the combinatorics operators). For example, the codecs parameter for an application/mp4 file that contains a TTML document that conforms to both IMSC1.0.1 text profile and EBU-TT-D-1 is "stpp.ttml.im1t|etd1".

It would be helpful to have some of the language above in this profiles document to clarify that this codecs parameter is not the same as the RFC 6381 codecs parameter.

**Comment by nigelmegitt** [comment]
> It would be helpful to have some of the language above in this profiles document to clarify that this codecs parameter is not the same as the RFC 6381 codecs parameter.

@mikedo is that not what you raised as #76? I _think_ this issue is about the formal definition of the use of the `|` and `+` operators.

**Comment by nigelmegitt** [comment]
> In particular, the proposal in bullet point 2 of [#63 (comment)](https://github.com/w3c/tt-profile-registry/issues/63#issuecomment-468380910) is a good text.

My understanding is that this issue is directly a request for a BNF definition of the syntax of the `codecs` parameter value. @cconcolato @mikedo it'd be good to have your confirmation that this would indeed resolve the issue, before a pull request gets raised.
